src/runtime/crdt/crdt-entity.ts

```diff
@@ -0,0 +1,202 @@
+/**
+ * @license
+ * Copyright (c) 2019 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {ChangeType, CRDTChange, CRDTModel, CRDTTypeRecord, VersionMap} from './crdt.js';
+import {Referenceable, CRDTCollectionTypeRecord, CRDTCollection, CollectionOpTypes} from './crdt-collection.js';
+import {CRDTSingletonTypeRecord, CRDTSingleton, SingletonOpTypes} from './crdt-singleton.js';
+import {Dictionary} from '../hot.js';
+
+type Identified = Dictionary<Referenceable>;
+
+type RawEntity<S extends Identified, C extends Identified> = 
+{
+  singletons: S,
+  collections: {[P in keyof C]: Set<C[P]>}
+};
+
+type SingletonEntityData<S extends Identified> = {[P in keyof S]: CRDTSingletonTypeRecord<S[P]>['data']};
+type CollectionEntityData<S extends Identified> = {[P in keyof S]: CRDTCollectionTypeRecord<S[P]>['data']};
+
+type SingletonEntityModel<S extends Identified> = {[P in keyof S]: CRDTSingleton<S[P]>};
+type CollectionEntityModel<S extends Identified> = {[P in keyof S]: CRDTCollection<S[P]>};
+
+export type EntityData<S extends Identified, C extends Identified> = 
+  {
+    singletons: SingletonEntityData<S>,
+    collections: CollectionEntityData<C>,
+    version: VersionMap
+  };
+
+export type EntityInternalModel<S extends Identified, C extends Identified> = 
+  {
+    singletons: SingletonEntityModel<S>,
+    collections: CollectionEntityModel<C>,
+    version: VersionMap
+  };
+
+export enum EntityOpTypes {Set, Clear, Add, Remove}
+
+type SetOp<S, F extends keyof S> = {type: EntityOpTypes.Set, field: F, value: S[F], actor: string, clock: VersionMap};
+type AddOp<C, F extends keyof C> = {type: EntityOpTypes.Add, field: F, value: C[F], actor: string, clock: VersionMap};
+type RemoveOp<C, F extends keyof C> = {type: EntityOpTypes.Remove, field: F, value: C[F], actor: string, clock: VersionMap};
+export type EntityOperation<S, C> = 
+  SetOp<S, keyof S> |
+  {type: EntityOpTypes.Clear, field: keyof S, actor: string, clock: VersionMap} |
```

Comment:
You don't want to define a ClearOp type?

Reply:
SetOp, etc. are only defined because the type system requires a separation from type description and instantiation (i.e. I need 2 type variables, one for the set and one for the field; I can't invent additional internal type variables at the SetOp declaration so I need this separation).